addons/marketing_campaign/models/marketing_campaign_segment.py

```diff
@@ -0,0 +1,144 @@
+        string='End Date', help="Date this segment was last closed or cancelled.")
+    date_next_sync = fields.Datetime(compute='_compute_next_sync', string='Next Synchronization',
+                                     help="Next time the synchronization job is scheduled to run automatically")
+
+    def _compute_next_sync(self):
+        # next auto sync date is same for all segments
+        sync_job = self.env.ref(
+            'marketing_campaign.ir_cron_marketing_campaign_every_day')
+        self.date_next_sync = sync_job.nextcall
+
+    @api.constrains('ir_filter_id', 'campaign_id')
+    def _check_model(self):
+        for obj in self:
+            if not self.ir_filter_id:
+                return True
+            if self.campaign_id.object_id.model != self.ir_filter_id.model_id:
+                raise ValueError(
+                    _('Model of filter must be same as resource model of Campaign.'))
+        return True
+
+    @api.onchange('campaign_id')
+    def onchange_campaign_id(self):
+        res = {'domain': {'ir_filter_id': []}}
+        if self.campaign_id:
+            campaign = self.campaign_id
+            if campaign.object_id:
+                mod_name = campaign.object_id.model
+                res['domain'] = {'ir_filter_id': [('model_id', '=', mod_name)]}
+        else:
+            self.ir_filter_id = False
+        return res
+
+    @api.multi
+    def state_running_set(self):
+        self.ensure_one()
+        vals = {'state': 'running'}
+        if not self.date_run:
+            vals['date_run'] = fields.Datetime.now()
+        return self.write(vals)
+
+    @api.multi
+    def state_done_set(self):
+        workitems = self.env['marketing.campaign.workitem'].search(
+            [('state', '=', 'todo'), ('segment_id', 'in', self.ids)])
+        workitems.write({'state': 'cancelled'})
+        return self.write(
+            {'state': 'done', 'date_done': fields.Datetime.now()})
+
+    @api.multi
+    def state_cancel_set(self):
+        workitems = self.env['marketing.campaign.workitem'].search(
+            [('state', '=', 'todo'), ('segment_id', 'in', self.ids)])
+        workitems.write({'state': 'cancelled'})
+        return self.write(
+            {'state': 'cancelled', 'date_done': fields.Datetime.now()})
+
+    @api.multi
+    def synchroniz(self):
+        return True if self.process_segment() else False
+
+    @api.model
+    def process_segment(self):
+        Workitem = self.env['marketing.campaign.workitem']
+        campaign = self.env['marketing.campaign']
+        segments = self.search([('state', '=', 'running')])
+
+        action_date = fields.Datetime.now()
+        for segment in segments.filtered(lambda x: x.campaign_id.state == 'running'):
+            campaign |= segment.campaign_id
+            activities = segment.campaign_id.activity_ids.filtered(lambda activity: (
+                activity.start == True and activity.campaign_id.id == segment.campaign_id.id))
```

Comment:
If we are filtering record from segment.campaign_id.activity_ids then do we need activity.campaign_id.id == segment.campaign_id.id ?